lib/Jsrt/JsrtDiag.cpp

```diff
@@ -0,0 +1,688 @@
+#include "JsrtDebugUtils.h"
+
+#define VALIDATE_IS_DEBUGGING(jsrtDebugManager) \
+    if (jsrtDebugManager == nullptr || !jsrtDebugManager->IsDebugEventCallbackSet()) \
+    { \
+        return JsErrorDiagNotDebugging; \
+    }
+
+#define VALIDATE_RUNTIME_IS_AT_BREAK(runtime) \
+    if (runtime->GetThreadContext()->GetDebugManager() == nullptr || !runtime->GetThreadContext()->GetDebugManager()->IsAtDispatchHalt()) \
+    { \
+        return JsErrorDiagNotAtBreak; \
+    }
+
+CHAKRA_API JsDiagStartDebugging(
+    _In_ JsRuntimeHandle runtimeHandle,
+    _In_ JsDiagDebugEventCallback debugEventCallback,
+    _In_opt_ void* callbackState)
+{
+    return GlobalAPIWrapper([&]() -> JsErrorCode {
+
+        VALIDATE_INCOMING_RUNTIME_HANDLE(runtimeHandle);
+
+        PARAM_NOT_NULL(debugEventCallback);
+
+        JsrtRuntime * runtime = JsrtRuntime::FromHandle(runtimeHandle);
+        ThreadContext * threadContext = runtime->GetThreadContext();
+
+        if (threadContext->GetRecycler() && threadContext->GetRecycler()->IsHeapEnumInProgress())
+        {
+            return JsErrorHeapEnumInProgress;
+        }
+        else if (threadContext->IsInThreadServiceCallback())
+        {
+            return JsErrorInThreadServiceCallback;
+        }
+        else if (threadContext->IsInScript())
+        {
+            return JsErrorRuntimeInUse;
+        }
+        else if (runtime->GetJsrtDebugManager() != nullptr && runtime->GetJsrtDebugManager()->IsDebugEventCallbackSet())
+        {
+            return JsErrorDiagAlreadyInDebugMode;
+        }
+
+        ThreadContextScope scope(threadContext);
+
+        if (!scope.IsValid())
+        {
+            return JsErrorWrongThread;
+        }
+
+        // Create the debug object to save callback function and data
+        runtime->EnsureJsrtDebugManager();
+
+        JsrtDebugManager* jsrtDebugManager = runtime->GetJsrtDebugManager();
+
+        jsrtDebugManager->SetDebugEventCallback(debugEventCallback, callbackState);
+
+        if (threadContext->GetDebugManager() != nullptr)
+        {
+            threadContext->GetDebugManager()->SetLocalsDisplayFlags(Js::DebugManager::LocalsDisplayFlags::LocalsDisplayFlags_NoGroupMethods);
+        }
+
+        for (Js::ScriptContext *scriptContext = runtime->GetThreadContext()->GetScriptContextList();
+        scriptContext != nullptr && !scriptContext->IsClosed();
+            scriptContext = scriptContext->next)
+        {
+            Assert(!scriptContext->IsScriptContextInDebugMode());
+
+            Js::DebugContext* debugContext = scriptContext->GetDebugContext();
+
+            if (debugContext->GetHostDebugContext() == nullptr)
+            {
+                debugContext->SetHostDebugContext(jsrtDebugManager);
+            }
+
+            if (FAILED(scriptContext->OnDebuggerAttached()))
+            {
+                AssertMsg(false, "Failed to start debugging");
+                return JsErrorFatal; // Inconsistent state, we can't continue from here?
```

Comment:
used Debugger_AttachDetach_fatal_error

---
In reply to: [62730641](https://github.com/Microsoft/ChakraCore/pull/926#discussion_r62730641) [](ancestors = 62730641)